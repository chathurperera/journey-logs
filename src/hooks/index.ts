export * from './useFirestorePagination';
export * from './useFetch';
