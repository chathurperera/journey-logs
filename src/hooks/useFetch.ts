import { useCallback, useEffect, useRef, useState } from 'react';

export const useFetch = fetchMethod => {
  const isMounted = useRef(true);

  const [state, setState] = useState({
    data: null,
    isLoading: false,
    isSuccess: false,
    isError: false,
  });

  const fetchData = useCallback(async () => {
    try {
      setState(prevState => ({
        ...prevState,
        isLoading: true,
        isSuccess: false,
        isError: false,
      }));

      const data = await fetchMethod();

      setState({
        data,
        isLoading: false,
        isSuccess: true,
        isError: false,
      });
    } catch (error) {
      setState({
        data: undefined,
        isLoading: false,
        isSuccess: false,
        isError: true,
      });
    }
  }, [fetchMethod]);

  useEffect(() => {
    fetchData();

    return () => {
      isMounted.current = false;
    };
  }, []);

  return { ...state };
};
