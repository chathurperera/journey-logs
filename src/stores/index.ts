import { RematchDispatch, RematchRootState, init } from '@rematch/core';
import loadingPlugin from '@rematch/loading';
import createRematchPersist from '@rematch/persist';
import {
  TypedUseSelectorHook,
  useDispatch as useReduxDispatch,
  useSelector as useReduxSelector,
} from 'react-redux';

import { IS_JEST_RUNTIME } from '@jl/constants';
import { PersistentStorageService } from '@jl/services';

import { RootModel, models } from './models';

export let store: ReturnType<typeof initializeStore>;

export const initializeStore = () => {
  const plugins = [];
  plugins.push(loadingPlugin());

  const persistPlugin = createRematchPersist({
    key: 'root',
    storage: PersistentStorageService.getStorage(),
    whitelist: ['userStore'],
  });

  if (!IS_JEST_RUNTIME) {
    plugins.push(persistPlugin);
  }

  const initializedStore = init({
    models,
    plugins,
  });

  store = initializedStore;

  return initializedStore;
};

export type Store = typeof store;
export type Dispatch = RematchDispatch<RootModel>;
export type RootState = RematchRootState<RootModel>;

export const useDispatch = () => useReduxDispatch<Dispatch>();
export const useSelector: TypedUseSelectorHook<RootState> = useReduxSelector;
export * from './models';
