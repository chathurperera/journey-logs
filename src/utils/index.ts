export * from './moment-utils';
export * from './render-utils';
export * from './PIN-utils';
