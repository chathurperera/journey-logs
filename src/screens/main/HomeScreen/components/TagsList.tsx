import { Icon } from '@rneui/base';
import React, { useCallback, useEffect, useState } from 'react';
import { FlatList, Pressable } from 'react-native';

import { Text } from '@jl/components';
import { tw } from '@jl/config';
import { Color, TextVariant } from '@jl/constants';
import { useFetch } from '@jl/hooks';
import { TagsService } from '@jl/services';
import { useSelector } from '@jl/stores';

type ItemProps = {
  item: string;
  onPress: (item: string) => void; // Use item as a parameter
  backgroundColor: string;
  textColor: string;
};

interface TagsListProps {
  selectedTag: string;
  setSelectedTag: React.Dispatch<React.SetStateAction<string>>;
}

const Item = React.memo(({ item, onPress, backgroundColor, textColor }: ItemProps) => (
  <Pressable
    onPress={() => onPress(item)}
    style={[
      tw`py-1.5 px-4 bg-black border mr-2 border-gray-200 rounded-[15px] gap-1 flex-row items-center`,
      { backgroundColor: backgroundColor },
    ]}>
    <Icon type="feather" name="hash" size={18} color={textColor} />
    <Text variant={TextVariant.Label2SemiBold} color={textColor}>
      {item}
    </Text>
  </Pressable>
));

export function TagsList({ selectedTag, setSelectedTag }: TagsListProps) {
  const { userId } = useSelector(state => state.userStore.userData);
  const { data = [] } = useFetch(() => TagsService.getAllTags(userId));
  const [allTags, setAllTags] = useState(['All']);

  const handlePress = useCallback(
    (item: string) => {
      setSelectedTag(item);
    },
    [setSelectedTag],
  );

  const renderItem = useCallback(
    ({ item }: { item: string }) => {
      const backgroundColor = item === selectedTag ? Color.Neutral.black : Color.Tertiary.JL200;
      const color = item === selectedTag ? Color.Neutral.white : Color.Neutral.JL900;

      return <Item item={item} onPress={handlePress} backgroundColor={backgroundColor} textColor={color} />;
    },
    [selectedTag],
  );

  useEffect(() => {
    console.log('data', data);
    if (data && data.length) {
      setAllTags(['All', ...data]);
    }
  }, [data]);

  return (
    <FlatList
      horizontal
      data={allTags}
      showsHorizontalScrollIndicator={false}
      renderItem={renderItem}
      keyExtractor={item => item}
    />
  );
}
