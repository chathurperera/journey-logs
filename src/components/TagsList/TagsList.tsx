import { Icon } from '@rneui/base';
import React, { useCallback } from 'react';
import { FlatList, Pressable, View } from 'react-native';

import { Text } from '@jl/components';
import { tw } from '@jl/config';
import { Color, TextVariant } from '@jl/constants';
import { useFetch } from '@jl/hooks';
import { TagsService } from '@jl/services';
import { useSelector } from '@jl/stores';

type ItemProps = {
  item: string;
  handleOnPress: (item: string) => void;
  backgroundColor: string;
  textColor: string;
};

interface TagsListProps {
  setSelectedTags: React.Dispatch<React.SetStateAction<string[]>>;
  selectedTags: string[];
  isEditable?: boolean;
  noteTags?: string[];
}

const Item = React.memo(({ item, handleOnPress, backgroundColor, textColor }: ItemProps) => {
  return (
    <Pressable
      onPress={() => handleOnPress(item)}
      style={[
        tw`py-2 px-4 bg-black border mr-2 border-gray-200 rounded-[20px] gap-1 flex-row items-center justify-between`,
        { backgroundColor: backgroundColor },
      ]}>
      <Icon type="feather" name="hash" size={18} color={textColor} />
      <Text variant={TextVariant.Body2Regular} color={textColor}>
        {item}
      </Text>
    </Pressable>
  );
});

export function TagsList({ setSelectedTags, selectedTags, isEditable = false, noteTags = [] }: TagsListProps) {
  const { userId } = useSelector(state => state.userStore.userData);
  const { data } = useFetch(() => TagsService.getAllTags(userId));

  const tagsToDisplay = isEditable ? data : noteTags;

  const toggleTag = useCallback(
    (tag: string) => {
      if (!isEditable) return;

      if (selectedTags.includes(tag)) {
        setSelectedTags((prevTags: string[]) => prevTags.filter((t: string) => t !== tag));
      } else {
        setSelectedTags((prevTags: string[]) => [...prevTags, tag]);
      }
    },
    [selectedTags, setSelectedTags, isEditable],
  );

  const renderItem = useCallback(
    ({ item }) => {
      const backgroundColor = selectedTags.includes(item) ? Color.Neutral.black : Color.Tertiary.JL200;
      const color = selectedTags.includes(item) ? Color.Neutral.white : Color.Neutral.JL900;
      return <Item item={item} handleOnPress={toggleTag} backgroundColor={backgroundColor} textColor={color} />;
    },
    [selectedTags, toggleTag],
  );

  return (
    <View style={tw`my-3`}>
      <FlatList
        horizontal
        contentContainerStyle={tw``}
        data={tagsToDisplay}
        showsHorizontalScrollIndicator={false}
        renderItem={renderItem}
        keyExtractor={item => item}
      />
    </View>
  );
}
