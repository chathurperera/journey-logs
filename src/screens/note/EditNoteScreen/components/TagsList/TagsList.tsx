import { Icon } from '@rneui/base';
import React, { useCallback } from 'react';
import { FlatList, Pressable, View } from 'react-native';

import { Text } from '@jl/components';
import { tw } from '@jl/config';
import { Color, TextVariant } from '@jl/constants';

type ItemProps = {
  item: string;
  handleOnPress: (item: string) => void;
  backgroundColor: string;
  textColor: string;
};

type TagsListProps = {
  setNoteTags: React.Dispatch<React.SetStateAction<string[]>>;
  noteTags: string[];
  allTags: string[];
};

const Item = React.memo(({ item, handleOnPress, backgroundColor, textColor }: ItemProps) => {
  return (
    <Pressable
      onPress={() => handleOnPress(item)}
      style={[
        tw`py-2 px-4 bg-black border mr-2 border-gray-200 rounded-[20px] gap-1 flex-row items-center justify-between`,
        { backgroundColor: backgroundColor },
      ]}>
      <Icon type="feather" name="hash" size={18} color={textColor} />
      <Text variant={TextVariant.Body2Regular} color={textColor}>
        {item}
      </Text>
    </Pressable>
  );
});

export function TagsList({ setNoteTags, noteTags, allTags }: TagsListProps) {
  const toggleTag = useCallback(
    (tag: string) => {
      if (noteTags?.includes(tag)) {
        setNoteTags(prevTags => prevTags.filter(t => t !== tag));
      } else {
        setNoteTags(prevTags => [...prevTags, tag]);
      }
    },
    [noteTags, setNoteTags],
  );

  const renderItem = useCallback(
    ({ item }) => {
      const isSelected = noteTags?.includes(item);
      return (
        <Item
          item={item}
          handleOnPress={toggleTag}
          backgroundColor={isSelected ? Color.Neutral.black : Color.Tertiary.JL200}
          textColor={isSelected ? Color.Neutral.white : Color.Neutral.JL900}
        />
      );
    },
    [noteTags, toggleTag],
  );

  return (
    <View style={tw`my-3`}>
      <FlatList
        horizontal
        contentContainerStyle={tw``}
        data={allTags}
        showsHorizontalScrollIndicator={false}
        renderItem={renderItem}
        keyExtractor={item => item}
      />
    </View>
  );
}
