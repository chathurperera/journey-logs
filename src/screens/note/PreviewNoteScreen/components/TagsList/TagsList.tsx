import { Icon } from '@rneui/base';
import React, { useCallback } from 'react';
import { FlatList, View } from 'react-native';

import { Text } from '@jl/components';
import { tw } from '@jl/config';
import { Color, TextVariant } from '@jl/constants';

interface TagsListProps {
  tags: string[];
}

export function TagsList({ tags }: TagsListProps) {
  console.log('tags in TgasList', tags);

  const renderItem = useCallback(
    ({ item }) => {
      console.log('item', item);
      return (
        <View
          style={tw`py-2 px-4 bg-black border mr-2 border-gray-200 rounded-[20px] gap-1 flex-row items-center justify-between bg-[${Color.Neutral.black}]`}>
          <Icon type="feather" name="hash" size={18} color={Color.Neutral.white} />
          <Text variant={TextVariant.Body2Regular} color={Color.Neutral.white}>
            {item}
          </Text>
        </View>
      );
    },
    [tags],
  );

  return (
    <View style={tw`my-3`}>
      <FlatList
        horizontal
        data={tags}
        showsHorizontalScrollIndicator={false}
        renderItem={renderItem}
        keyExtractor={item => item}
      />
    </View>
  );
}
