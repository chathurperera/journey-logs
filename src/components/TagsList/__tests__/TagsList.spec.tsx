import React from 'react';

import { renderWithProviders } from '@jl/utils';

import { TagsList } from '../TagsList';

jest.mock('@jl/hooks', () => ({
  useFetch: jest.fn(),
}));

jest.mock('@jl/stores', () => ({
  useSelector: jest.fn(),
}));

describe('Describe <TagsList />', () => {
  const mockSetSelectedTags = jest.fn();
  const testID = 'TagsListTestID';

  const getRenderedComponent = () =>
    renderWithProviders(<TagsList selectedTags={['All']} setSelectedTags={mockSetSelectedTags} testID={testID} />);

  beforeEach(() => {
    jest.resetAllMocks();

    require('@jl/hooks').useFetch.mockReturnValue({
      data: ['tag1', 'tag2'],
    });

    require('@jl/stores').useSelector.mockReturnValue({
      userStore: {
        userData: {
          userId: '12345',
        },
      },
    });
  });

  it('should find the LoginScreen via testID', () => {
    const { getByTestId } = getRenderedComponent();
    const foundScreen = getByTestId(testID);
    expect(foundScreen).toBeTruthy();
  });

  it('should render TagsList correctly', () => {
    const renderer = getRenderedComponent();
    const renderTree = renderer.toJSON();
    expect(renderTree).toMatchSnapshot();
  });

  it('renders tags when isEditable is false', () => {
    const { getByText } = renderWithProviders(
      <TagsList selectedTags={['All']} setSelectedTags={mockSetSelectedTags} isEditable={false} />,
    );

    expect(getByText('All')).toBeTruthy();
    // The component should not render 'tag1' and 'tag2' when isEditable is false
    expect(() => getByText('tag1')).toThrow();
    expect(() => getByText('tag2')).toThrow();
  });
});
