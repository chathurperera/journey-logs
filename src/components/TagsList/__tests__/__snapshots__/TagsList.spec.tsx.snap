// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Describe <TagsList /> should render TagsList correctly 1`] = `
<View
  style={
    {
      "marginBottom": 12,
      "marginTop": 12,
    }
  }
  testID="TagsListTestID"
>
  <RCTScrollView
    contentContainerStyle={{}}
    data={
      [
        "All",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "#000",
                "borderColor": "#e5e7eb",
                "borderRadius": 20,
                "borderWidth": 1,
                "flexDirection": "row",
                "gap": 4,
                "justifyContent": "space-between",
                "marginRight": 8,
                "paddingBottom": 8,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 8,
              },
              {
                "backgroundColor": "#000000",
              },
            ]
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#ffffff",
                        "fontSize": 18,
                      },
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "fontFamily": "Feather",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  ï…¿
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              [
                {
                  "color": "#ffffff",
                },
                {
                  "fontFamily": "Inter-Regular",
                  "fontSize": 18,
                  "fontWeight": "normal",
                  "textAlign": "left",
                  "textTransform": "none",
                },
              ]
            }
          >
            All
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
